File: AoC/2023_problems/day8.py
import math


with open("AoC/2023_problems/day8.txt") as f:
    data = f.read()


def lcm(nums):
    lcm = 1
    for num in nums:
        lcm = lcm * num // math.gcd(lcm, num)
    return lcm


def solve(text):
    instructions, nodes = text.split("\n\n")
    node_dict = {}
    for node in nodes.split("\n"):
        name, children = node.split(" = (")
        children = children.replace(")", "").split(", ")
        node_dict[name] = children
    current_node = "AAA"
    i, p1 = 0, 0
    while current_node != "ZZZ":
        if i == len(instructions):
            i = 0
        next_step = 0 if instructions[i] == "L" else 1
        current_node = node_dict[current_node][next_step]
        i += 1
        p1 += 1
    # part 2
    p2 = lcm(solve_p2(instructions, node_dict))
    print(p1, p2)


def solve_p2(instructions,  node_dict):
    current_nodes = list(filter(lambda x: x[-1] == "A", node_dict.keys()))
    substeps = []
    while current_nodes:
        node = current_nodes.pop()
        substeps.append(0)
        i = 0
        while node[-1] != "Z":
            if i == len(instructions):
                i = 0
            next_step = 0 if instructions[i] == "L" else 1
            node = node_dict[node][next_step]
            substeps[-1] += 1
            i += 1
    return substeps

solve(data)

File: AoC/2023_problems/day8.js
const fs = require("fs");

const data = fs.readFileSync("day8.txt", "utf8")

function parseData(data) {
    // Extract moves as string and nodes as {node: [left, right]}
    const [moves, nodes] = data.split("\n\n");
    const nodesGraph = {};
    for (const line of nodes.split("\n")) {
        const [node, children] = line.split(" = ");
        const leftChild = children.slice(1).split(", ")[0];
        const rightChild = children.split(", ")[1].slice(0, -1);
        nodesGraph[node] = [leftChild, rightChild];
    }
    return [moves, nodesGraph];
}

function gcd(a, b) {
    if (b === 0) {
        return a;
    }
    return gcd(b, a % b);
}

function lcm(nums) {
    let lcmVal = 1;
    for (let num of nums) {
        lcmVal = lcmVal * num / gcd(lcmVal, num);
    }
    return lcmVal;
}

function solve(data) {
    const [moves, nodesGraph] = parseData(data);
    // Extract all nodes ending with A
    const aNodes = Object.keys(nodesGraph).filter(node => node.endsWith("A"));
    let nodesSteps = {};
    
    // Take aNodes one by one and cycle until every one gets to Z node
    for (let node of aNodes) {
        nodesSteps[node] = 0;
        let currentNode = node;

        let i = 0;
        while (!currentNode.endsWith("Z")) {
            if (i == moves.length) i = 0;
            let move = moves[i] == "L" ? 0 : 1;
            currentNode = nodesGraph[currentNode][move];
            nodesSteps[node]++;
            i++;
        }
    }
    return [nodesSteps["AAA"], lcm(Object.values(nodesSteps))];
}

console.log(solve(data));
